### PR TITLE
feat(cart): add cart item entities state

### PR DESCRIPTION
**Pull request by lderrickable**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/graycoreio/daffodil/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information
Plan for the future is to remove the cart items state from the cart state (it would be a Partial<T extends DaffCart>. When the cart is requested through a selector, add the cart item state into the cart as it is being returned to the user.

### CODE REVIEW

File: libs/cart/src/selectors/cart-order/cart-order.selector.ts [hunk 1]
```diff
@@ -22,13 +23,14 @@ export interface DaffCartOrderMemoizedSelectors<
 
 const createCartOrderSelectors = <
   T extends DaffCart = DaffCart,
-  V extends DaffCartOrderResult = DaffCartOrderResult
+	V extends DaffCartOrderResult = DaffCartOrderResult,
+	U extends DaffCartItem = DaffCartItem
 >(): DaffCartOrderMemoizedSelectors<V> => {
 	const selectCartFeatureState = getDaffCartFeatureSelector<T, V>().selectCartFeatureState;
```
Comment (griest024) [inline libs/cart/src/selectors/cart-order/cart-order.selector.ts:29]: Pass generic into `getDaffCartFeatureSelector`

File: libs/cart/src/selectors/cart.selector.ts [hunk 2]
```diff
@@ -3,32 +3,39 @@ import { DaffCartOrderResult } from '../models/cart-order-result';
...
-  V extends DaffCartOrderResult = DaffCartOrderResult
+	V extends DaffCartOrderResult = DaffCartOrderResult,
+	U extends DaffCartItem = DaffCartItem
 >(): DaffCartMemoizedSelectors<T> => {
 
 	return {
 		...getDaffCartFeatureSelector<T, V>(),
 		...getCartOrderSelectors<T, V>(),
-		...getCartSelectors<T>()
+		...getCartSelectors<T>(),
```
Comment (griest024) [inline libs/cart/src/selectors/cart.selector.ts:27]: Pass generics into the selector getters.

File: libs/cart/src/selectors/cart-order/cart-order.selector.ts [hunk 3]
```diff
@@ -22,13 +23,14 @@ export interface DaffCartOrderMemoizedSelectors<
 
 const createCartOrderSelectors = <
   T extends DaffCart = DaffCart,
-  V extends DaffCartOrderResult = DaffCartOrderResult
+	V extends DaffCartOrderResult = DaffCartOrderResult,
+	U extends DaffCartItem = DaffCartItem
```
Comment (griest024) [inline libs/cart/src/selectors/cart-order/cart-order.selector.ts:27]: You need to add this generic to `getCartOrderSelectors` as well.

File: libs/cart/src/selectors/cart/cart.selector.ts [hunk 4]
```diff
@@ -62,12 +62,13 @@ export interface DaffCartStateMemoizedSelectors<
 
 const createCartSelectors = <
   T extends DaffCart = DaffCart,
-  V extends DaffCartOrderResult = DaffCartOrderResult
+	V extends DaffCartOrderResult = DaffCartOrderResult,
+	U extends DaffCartItem = DaffCartItem
```
Comment (griest024) [inline libs/cart/src/selectors/cart/cart.selector.ts:66]: Add this generic to `getCartSelectors` as well.

File: libs/cart/src/facades/cart/cart-facade.interface.ts [hunk 5]
```diff
@@ -30,6 +33,7 @@ export interface DaffCartFacadeInterface<
   grandTotal$: Observable<DaffCart['grand_total']>;
   coupons$: Observable<DaffCart['coupons']>;
   items$: Observable<DaffCart['items']>;
+  dictionaryOfItems$: Observable<Dictionary<U>>;
```
Comment (griest024) [inline libs/cart/src/facades/cart/cart-facade.interface.ts:36]: I think `itemDictionary$` is a better name.

File: libs/cart/src/reducers/cart-item-entities/cart-item-entities-reducer-adapter.ts [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+
+import { DaffCartItem } from '../../models/cart-item';
+
+/**
+ * Cart Item Entities Adapter for changing/overwriting entity state.
+ */
+export const daffCartItemEntitiesAdapter = (() => {
+	let cache;
+  return <T extends DaffCartItem>(): EntityAdapter<T> =>
+    cache = cache || createEntityAdapter<T>();
```
[the remark this answers is not in the document]
Comment (lderrickable) [reply libs/cart/src/reducers/cart-item-entities/cart-item-entities-reducer-adapter.ts]: oh nice. Changed

File: libs/cart/src/selectors/cart-order/cart-order.selector.ts (hunk 1 above)
Comment (lderrickable) [reply libs/cart/src/selectors/cart-order/cart-order.selector.ts]: done

File: libs/cart/src/selectors/cart.selector.ts (hunk 2 above)
Comment (lderrickable) [reply libs/cart/src/selectors/cart.selector.ts]: done

File: libs/cart/src/selectors/cart-order/cart-order.selector.ts (hunk 3 above)
Comment (lderrickable) [reply libs/cart/src/selectors/cart-order/cart-order.selector.ts:27]: done

File: libs/cart/src/selectors/cart/cart.selector.ts (hunk 4 above)
Comment (lderrickable) [reply libs/cart/src/selectors/cart/cart.selector.ts:66]: done

File: libs/cart/src/facades/cart/cart-facade.interface.ts (hunk 5 above)
Comment (lderrickable) [reply libs/cart/src/facades/cart/cart-facade.interface.ts]: agreed. changed

File: libs/cart/src/selectors/cart/cart.selector.ts [hunk 7]
```diff
@@ -62,12 +62,13 @@ export interface DaffCartStateMemoizedSelectors<
...
 >(): DaffCartStateMemoizedSelectors<T> => {
 	const selectCartFeatureState = getDaffCartFeatureSelector<T, V>().selectCartFeatureState;
```
Comment (griest024) [inline libs/cart/src/selectors/cart/cart.selector.ts:68]: Pass generic into `getDaffCartFeatureSelector`

Comment (lderrickable) [reply libs/cart/src/selectors/cart/cart.selector.ts]: good eye. fixed

**Pull request merged**